pkg/controllers/requirement/reconciler.go

```diff
@@ -0,0 +1,219 @@
+	errGetSecret               = "cannot get secret"
+	errApplySecret             = "cannot apply secret"
+)
+
+// Event reasons.
+const (
+	reasonCannotGetFromRemote   event.Reason = "CannotGetFromRemote"
+	reasonCannotAddFinalizer    event.Reason = "CannotAddFinalizer"
+	reasonCannotRemoveFinalizer event.Reason = "CannotRemoveFinalizer"
+	reasonCannotPropagate       event.Reason = "CannotPropagate"
+	reasonCannotDelete          event.Reason = "CannotDelete"
+)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(rec event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = rec
+	}
+}
+
+// WithFinalizer specifies how the Reconciler should add and remove finalizers.
+func WithFinalizer(f rresource.Finalizer) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.finalizer = f
+	}
+}
+
+// WithPropagator specifies how the Reconciler should propagate values and objects
+// between clusters.
+func WithPropagator(p Propagator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.requirement = p
+	}
+}
+
+// ReconcilerOption is used to configure *Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// NewReconciler returns a new *Reconciler.
+func NewReconciler(mgr manager.Manager, remoteClient client.Client, gvk schema.GroupVersionKind, opts ...ReconcilerOption) *Reconciler {
+	ni := func() *requirement.Unstructured { return requirement.New(requirement.WithGroupVersionKind(gvk)) }
+	lc := unstructured.NewClient(mgr.GetClient())
+	lca := rresource.ClientApplicator{
+		Client:     lc,
+		Applicator: rresource.NewAPIPatchingApplicator(lc),
+	}
+	rc := unstructured.NewClient(remoteClient)
+	rca := rresource.ClientApplicator{
+		Client:     rc,
+		Applicator: rresource.NewAPIPatchingApplicator(rc),
+	}
+	r := &Reconciler{
+		mgr:         mgr,
+		local:       lca,
+		remote:      rca,
+		newInstance: ni,
+		log:         logging.NewNopLogger(),
+		finalizer:   rresource.NewAPIFinalizer(lc, finalizer),
+		requirement: NewPropagatorChain(
+			NewSpecPropagator(rca),
+			NewLateInitializer(lc),
+			NewStatusPropagator(),
+			NewConnectionSecretPropagator(lca, rca),
+		),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// Propagator is used to propagate values between objects.
+type Propagator interface {
+	Propagate(ctx context.Context, local, remote *requirement.Unstructured) error
+}
+
+// Reconciler syncs the given requirement instance from local cluster to remote
+// cluster and fetches its connection secret to local cluster if it's available.
+type Reconciler struct {
+	mgr    ctrl.Manager
+	local  rresource.ClientApplicator
+	remote rresource.ClientApplicator
+
+	newInstance func() *requirement.Unstructured
+
+	finalizer   rresource.Finalizer
+	requirement Propagator
```

Comment:
I found the name `requirement` a little confusing, because the name looks like it could represent a requirement resource, not a thing that knows how to perform `Propagate` for the resource. Elsewhere in this review I'm urging you to consider breaking this up, so it's a moot point if you decide to act on that feedback.